### ISSUE TITLE
PR merging automation

### ISSUE THREAD

**Issue by soullivaneuh** [sonata-project/dev-kit#81]
It would be great to have our PR merging automation waiting https://github.com/isaacs/github/issues/586.

Obviously, it's a dangerous part and should be done with care. Some important points/tasks:
- DevKit hook have to be set on **all** sonata repositores -> https://github.com/sonata-project/dev-kit/pull/83
- A PR can be merged only if the `RTM` label is set
- We have to check if PR does not have conflicts before call github API for merge
- SonataCI should comment the `RTM` PR to tell this one will be automatically merge
- Permit every user (even not contributors) to cancel the `RTM` trough a comment message like "Don't merge this PR[.!]"
- Add a doc for contributors of "How to use the RTM" label (PR title, commits check etc...)

The SonataCI automatic merge message:

```
This PR will be automatically merged.

If you have to do some modifications before, answer with `Don't merge this PR` message or simply remove the `RTM` tag yourself if you can.
```

Two ways for automatic merging:
- On status event : https://developer.github.com/v3/activity/events/types/#statusevent
  Take care to check the green status is corresponding to the **last** commit of the PR
- ~~On a cron task, fetching `RTM` PR and merge them.~~

~~Maybe both could be implemented if no conflicts.~~

-> Cron task is a bad idea because if GitHub hooks are down, we might merge PR that should not be merged.

Anything else I perhaps forgot?

**Comment by greg0ire** [comment]
> We have to check if PR does not have conflicts before call github API for merge

Do we? I think it is impossible to merge if there are conflicts.

**Comment by Soullivaneuh** [comment]
@greg0ire Precision: Ask github if we can merge this PR before call the merge API.

Otherwise, we might have request exceptions.

But of course the PR would not be merged even if we ask GitHub for it.

**Comment by Th3Mouk** [comment]
I don't know, auto merging is a bit risky if we totally remove human verification.

What about if the person add mess in documentation, push malicious code not covered by tests, remove all the tests etc...

IMO the automation can be done, but missing a step between the bots merge and the `RTM` label.

What about a `LGTM` comment by a sonata/contributors ? If the bot receive this comment and the PR has a `RTM` label then bots merge the PR.

The `don't merge this PR` message is a really good idea, but it's necessary to let the cron few minutes, by adding a date check on last commit.

**Comment by Soullivaneuh** [comment]
> auto merging is a bit risky if we totally remove human verification.

Human verification is not removed. The `RTM` must be set **only** if the PR is ready to merge (content plus commits).

> What about if the person add mess in documentation, push malicious code not covered by tests, remove all the tests etc...

If a commit is added, modified, rebased or whatever, the `RTM` label is automatically removed by SonataCI. See #80.

> What about a `LGTM` comment by a sonata/contributors ? If the bot receive this comment and the PR has a `RTM` label then bots merge the PR.

The `LGTM` comment has not the same goal IMHO. If I say LGTM on a PR, that would say "I thinks it's ok for me.". But another contributors would maybe like to add some asking before merge.

Set the `RTM` label is like pushing the merge button, but with CI waiting.

> The don't merge this PR message is a really good idea, but it's necessary to let the cron few minutes, by adding a date check on last commit.

Can't be possible because PR merging will be based on the PR status hooks event.

But as I said before, `RTM` would be the same as merge button click. So the user have a "last chance" to cancel it but if it's already merged and the user want to add something else, he can other another one. :wink: 

Does this answer to your worries? Feel free to elaborate! :+1:

**Comment by Soullivaneuh** [comment]
BTW, in any case if an error is merged despite of all the checks, it can always be reverted.

**Comment by core23** [comment]
Maybe we can use the Github social icons for this? When the build is green and a PR reaches 10(?) 👍 icons and not any 👎 icon we could merge this automatically.

**Comment by greg0ire** [comment]
> Maybe we can use the Github social icons for this? When the build is green and a PR reaches 10(?) :+1: icons and not any :-1: icon we could merge this automatically.

3 :+1: **from sonata members** would be best I think, and they should be put on the very first message.

**Comment by Soullivaneuh** [comment]
As I said. Put a `RTM` is the same as pressing the merge button. So we should consider this before using it.

> 3 :+1: **from sonata members** would be best I think, and they should be put on the very first message.

I'm not sure we can see which people put a :+1: with the GitHub api.

**Comment by Th3Mouk** [comment]
I wanted to propose this principe but i don't, because i think everybody can 👍 not only sonata contributor.
And the opposite case, if the limit is 3 👍 and 8 contributors are not present during the month, the automation will not work.

**Comment by core23** [comment]
> I wanted to propose this principe but i don't, because i think everybody can 👍 not only sonata contributor.

We could increase / reduce the number of upvotes, as there is not much traffic on GitHub.

> And the opposite case, if the limit is 3 👍 and 8 contributors are not present during the month, the automation will not work.

This is about auto merging. We can still merge this by hand.

**Comment by Th3Mouk** [comment]
So automerging shows no interrest if we continue to interact on github. Simply click merge on `RTM` PR

**Comment by Soullivaneuh** [comment]
> So automerging shows no interrest if we continue to interact on github. Simply click merge on `RTM` PR

Imagine a fresh simple PR adding a typo fixing on the doc. The PR is OK and you want to merge it. But you can't. Why? Because Travis has 42 jobs to do before this one and will not check this PR before tomorrow (Yes, Travis is a human and he sleeps, true story. :p).

Adding the `RTM` will allow dev-kit to do the merge automatically so you don't have to do it yourself.

It's just about saving time, not replacing reviewer job.

> What about if the person add mess in documentation, push malicious code not covered by tests, remove all the tests etc...

Another point about this: It can not be possible, thanks to this: https://developer.github.com/v3/pulls/#merge-a-pull-request-merge-button

> sha	string	SHA that pull request head must match to allow merge

In a nutshell, you have to pass to the api the sha which you think is the last commit of this PR.

If the PR is updated while you're performing the merge request though api, you will get this error response: https://developer.github.com/v3/pulls/#response-if-sha-was-provided-and-pull-request-head-did-not-match

In addition, if new commits are added, the `RTM` label will be removed by dev-kit: https://github.com/sonata-project/dev-kit/blob/5ce924800415066a44fc1971cae49d9192ce8b23/src/GithubHookProcessor.php#L81-L83

With that, I can ensure you that no malicious code can be added beside an automatic merge. :wink:

**Comment by VincentLanglet** [comment]
The feature is provided by github now
